### PR TITLE
System carousel theming

### PR DESCRIPTION
**Pull request by zigurana**
Hello dear RetroPie Core team!
I would like to offer some changes that allow the theming of the system carousel element, as shown in the system view of EmulationStation. This was developed together with @jrassa, who was a great help in finetuning the default behaviour, and ironing out some of the bugs.

This feature has been tested and all discovered bugs have been crushed, see the forum thread [here ](https://retropie.org.uk/forum/topic/2696/input-needed-es-theming-improvements/145).

As the development took a while, and was parallel to other developments on this stream, there were a couple of rebases I wanted to have in, in order to do better testing. As a result, there were many more commits than only mine on the branch, which were very tiresome to disentangle. So while I intended to squash all the commits into a couple/single commits, I offer it up as is. 
Luckily, since last year, the repo owner can simply press the "squash and merge" button, to create a single merge-commit into the master branch. Why this is not possible from my side, I have no idea, Git still baffles me often.

So, please have a look at the code, and let me know what you think.

In the meantime Sirs, I remain your most faithful and obedient servant.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -34,3 +34,6 @@ Makefile
 .cproject
 .project
 .settings/
+
+# WinMerge
+.bak
```
Comment (joolswills) [inline .gitignore]: This should not be included in the PR.

File: es-core/src/components/TextComponent.cpp [hunk 2]
```diff
@@ -73,32 +90,32 @@ void TextComponent::render(const Eigen::Affine3f& parentTrans)
...
 		trans = roundMatrix(trans);
 		Renderer::setMatrix(trans);
 
 		// draw the text area, where the text actually is going
-		if(Settings::getInstance()->getBool("DebugText"))
+		if (Settings::getInstance()->getBool("DebugText"))
 		{
-			switch(mAlignment)
+			switch (mAlignment)
 			{
```
Comment (joolswills) [inline es-core/src/components/TextComponent.cpp]: There is a bunch of whitespace reformatting in this file that should be done separately (or not done depending on other file styling).

File: es-app/src/views/ViewController.cpp [hunk 3]
```diff
@@ -12,7 +12,9 @@
 #include "animations/LaunchAnimation.h"
 #include "animations/MoveCameraAnimation.h"
 #include "animations/LambdaAnimation.h"
-#include <SDL2/SDL.h>
+//#include <SDL2/SDL.h>
```
Comment (joolswills) [inline es-app/src/views/ViewController.cpp]: This can be removed instead of commenting out of it is wrong.

**Comment by joolswills** [comment]
A few comments about formatting/whitespace.

There are a lot of commits here including merges etc which shouldn't be in the PR. Please can you rework/squash this down and try and remove whitespace changes, so the PR just contains changed code etc.

**Comment by joolswills** [comment]
Thanks for the work on this.

**Comment by zigurana** [comment]
Hi @joolswills, thanks for reviewing.

re. the whitespace spaces, I'll fix them.
re. the sdl2/sdl.h and gitignore, I'll fix those as well.

re. all the commits: 
As I had indicated, with the improvements on github.com last year, it is *much* easier for the repo owner to squash down everything when accepting the code changes, than it is for me. 
This is especially the case when developments take a bit longer, and when you want to merge in the changes from master on the dev stream in order to test before submitting the new changes. I feel that that should the preferred WoW, as it results in better tested PRs. But is has the result that there are always merge commits that de-facto empty commits from the masters point of view. Now, once you squash and merge, all these merge commits no longer matter, is only while considering the PR, they might complicate reviewing the code.
For instance, lets say I commit the whitespace fixes, would it be better if I rebased everything into a single commit, or better if you had just another additional commit (with a much smaller delta to consider/review)? Honest question, I simply do not know.

I tried to do an interactive rebase earlier, but it has the result that all the merge-related commits became committed (not authored) by me. This led to a very ugly PR as well.
Now, I can do a merge -squash myself, but that would need to happen on a new branch though, and will (I think) result in a new PR, and [it will do exactly the same as your button does](http://stackoverflow.com/questions/35948131/github-squash-commits-from-web-interface-on-pull-request-after-review-comments).

For future features, I will do an interactive rebase more often, to prevent a build-up of small commits like this (probably before each merge-in of master).

Please let me know how you want me to proceed.

Thanks again,

**Comment by joolswills** [comment]
I prefer to avoid the github gui function if possible. Also it doesn't allow for sensible grouped commits - eg with a feature commit, it might be good in some cases to group the commits into a few commits etc.

I would prefer a PR to be done once a branch is ready with the commits sorted. Any last minute fixes be manually squashed into the related commits (eg whitespace).

You don't need to do a new branch - you can reset/force push back to this branch when it is sorted.

**Comment by zigurana** [comment]
Ok, I'll see if I can manage. 
So just to be clear, you'd rather have a have nice handfull of commits, grouped by (sub) feature, than that I ruthlessly squash everything here down into a single commit?

**Comment by joolswills** [comment]
if they are logically grouped - eg fixes for previous commits should be squashed. I don't mind if it's all done as one also.

**Comment by zigurana** [comment]
Ok clear. Thanks for your time on this.

**Comment by joolswills** [comment]
Thanks for your contributions.

**Comment by zigurana** [comment]
Well, that was nerve-wrecking! but it seemed to have turned out OK. I have left it at three commits: 
1. The changes needed to compile with VS2015
3. The rebase to get in all the code that was changed during dev (not sure how to get rid of this one?)
2. The SystemCarousel feature itself.

**Comment by joolswills** [comment]
the second commit doesn't look like a rebase but a merge - you should be able to rebase your changes over the latest code without including a merge commit which I don't really want to include.

**Comment by joolswills** [comment]
eg https://github.com/RetroPie/EmulationStation/compare/master...joolswills:SystemCaroussel?expand=1

I made this by checking out your branch and doing `git rebase retropie/master` where retropie is the retropie repository.

**Comment by zigurana** [comment]
Done, thanks for holding my hand on this one. I'll try to do better next time!

**Comment by joolswills** [comment]
thanks - will give this a test when I have a chance. If anyone else has time to test that would be good too. As we are v.close to retropie 4.2 images, I want to avoid any issues - but im sure it is bug free ;-)

**Comment by jrassa** [comment]
There is a small one line change that is needed to make the carousel themeing compatible with the feature element for backward compatibility (#96).  I can submit a separate pull request if that is prefered.

https://github.com/jrassa/EmulationStation/commit/837dffdc70f3feace8cafff8df123a2aea874ac5

**Comment by joolswills** [comment]
Might be easier for testing if @zigurana can include that in this PR if that's ok ?

**Comment by zigurana** [comment]
Hey, I'm not sure when I can get to it, in the hospital atm (dad had a heart infarction). But feel free to do a PR on the contributing branch for this PR.

**Comment by joolswills** [comment]
Very sorry to hear that - my best wishes to your dad for a speedy and full recovery. NP at all regarding the change, can do it as a separate PR or just commit it after.

Cheers.

**Comment by zigurana** [comment]
@jrassa & @joolswills: I've amended the inclusion of carousel to the list of supported features in ThemeData.cpp to the last commit. 
Thanks for the reminder to include that.
@jrassa : I just did it by hand, rather than merge it in from your branch, so you're not mentioned explicitly. I just want to say, you hold a special place in my heart in any case ;-).

File: CMakeLists.txt [hunk 4]
```diff
@@ -81,9 +85,14 @@ if(CMAKE_COMPILER_IS_GNUCXX)
     endif()
 
     #set up compiler flags for GCC
+if (CMAKE_BUILD_TYPE MATCHES Debug)
+	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -Wno-attributes -O0") #support C++11 for std::, optimize
```
Comment (joolswills) [inline CMakeLists.txt]: check your indentation - there is a tab here (but the rest of the file uses spaces including your following line)

**Comment by joolswills** [comment]
everything looks good and it works fine for me on my pc. Once that rogue tab in CMakeLists.txt has been sorted, I will merge. thanks.

**Comment by zigurana** [comment]
Ok, changed the tab to 4 spaces.

**Comment by joolswills** [comment]
thanks.

**Comment by joolswills** [comment]
I missed this before, but this commit introduces a warning - 

```
/home/jools/Repos/RetroPie/EmulationStation/es-core/src/components/TextComponent.cpp: In constructor ‘TextComponent::TextComponent(Window*)’:
/home/jools/Repos/RetroPie/EmulationStation/es-core/src/components/TextComponent.cpp:11:163: warning: converting to non-pointer type ‘unsigned int’ from NULL [-Wconversion-null]
 se), mColor(0x000000FF), mAutoCalcExtent(true, true), mAlignment(ALIGN_LEFT), mLineSpacing(1.5f), mBgColor(NULL)
                                                                                                                ^
/home/jools/Repos/RetroPie/EmulationStation/es-core/src/components/TextComponent.cpp: In constructor ‘TextComponent::TextComponent(Window*, const string&, const std::shared_ptr<Font>&, unsigned int, Alignment, Eigen::Vector3f, Eigen::Vector2f, unsigned int)’:
/home/jools/Repos/RetroPie/EmulationStation/es-core/src/components/TextComponent.cpp:17:135: warning: converting to non-pointer type ‘unsigned int’ from NULL [-Wconversion-null]
 e(false), mColor(0x000000FF), mAutoCalcExtent(true, true), mAlignment(align), mLineSpacing(1.5f), mBgColor(NULL)
```

I assume it should be

```
diff --git a/es-core/src/components/TextComponent.cpp b/es-core/src/components/TextComponent.cpp
index 9ab5effa..c7b7a490 100644
--- a/es-core/src/components/TextComponent.cpp
+++ b/es-core/src/components/TextComponent.cpp
@@ -8,13 +8,13 @@
 #include "Settings.h"
 
 TextComponent::TextComponent(Window* window) : GuiComponent(window), 
-       mFont(Font::get(FONT_SIZE_MEDIUM)), mUppercase(false), mColor(0x000000FF), mAutoCalcExtent(true, true), mAlignment(ALIGN_LEFT), mLineSpacing(1.5f), mBgColor(NULL)
+       mFont(Font::get(FONT_SIZE_MEDIUM)), mUppercase(false), mColor(0x000000FF), mAutoCalcExtent(true, true), mAlignment(ALIGN_LEFT), mLineSpacing(1.5f), mBgColor(0)
 {
 }
 
 TextComponent::TextComponent(Window* window, const std::string& text, const std::shared_ptr<Font>& font, unsigned int color, Alignment align,
        Eigen::Vector3f pos, Eigen::Vector2f size, unsigned int bgcolor) : GuiComponent(window), 
-       mFont(NULL), mUppercase(false), mColor(0x000000FF), mAutoCalcExtent(true, true), mAlignment(align), mLineSpacing(1.5f), mBgColor(NULL)
+       mFont(NULL), mUppercase(false), mColor(0x000000FF), mAutoCalcExtent(true, true), mAlignment(align), mLineSpacing(1.5f), mBgColor(0)
 {
        setFont(font);
        setColor(color);
```

Will make the change.

**Comment by Supernature2k** [comment]
Hello everybody!

Regarding this enhancement, is there any plan to also change the slide transition?

i.e. when carousel is set to vertical, slide transition is up/down and not left/right

**Comment by zigurana** [comment]
I don't think that this is planned, but it sounds like a good idea. 
I have no idea how much work it will be,  though. Can you please open a new issue for this on the tracker? 
That way anyone who wants to get his/her hands dirty can get cracking!

**Comment by jrassa** [comment]
@Supernature2k No need to create an new issue.  I have a working implementation that I will submitting in the near future.
